internal/clients/compute/user/user.go

```diff
@@ -0,0 +1,103 @@
+package user
+
+import (
+	"context"
+
+	ionosdk "github.com/ionos-cloud/sdk-go/v6"
+
+	"github.com/ionos-cloud/crossplane-provider-ionoscloud/apis/compute/v1alpha1"
+	"github.com/ionos-cloud/crossplane-provider-ionoscloud/internal/clients"
+)
+
+// APIClient is a wrapper around IONOS Service
+type APIClient struct {
+	*clients.IonosServices
+}
+
+type Client interface {
+	GetUser(ctx context.Context, id string) (ionosdk.User, *ionosdk.APIResponse, error)
+	CreateUser(ctx context.Context, u ionosdk.UserPost) (ionosdk.User, *ionosdk.APIResponse, error)
+	UpdateUser(ctx context.Context, id string, u ionosdk.UserPut) (ionosdk.User, *ionosdk.APIResponse, error)
+	DeleteUser(ctx context.Context, id string) (*ionosdk.APIResponse, error)
+	GetAPIClient() *ionosdk.APIClient
+}
+
+// GetUser retrieves a user via its id.
+func (ac *APIClient) GetUser(ctx context.Context, id string) (ionosdk.User, *ionosdk.APIResponse, error) {
+	return ac.ComputeClient.UserManagementApi.UmUsersFindById(ctx, id).Execute()
+}
+
+// CreateUser creates a user in the ionoscloud.
+func (ac *APIClient) CreateUser(ctx context.Context, u ionosdk.UserPost) (ionosdk.User, *ionosdk.APIResponse, error) {
+	return ac.ComputeClient.UserManagementApi.UmUsersPost(ctx).User(u).Execute()
+}
+
+// UpdateUser updates a user.
+func (ac *APIClient) UpdateUser(ctx context.Context, id string, u ionosdk.UserPut) (ionosdk.User, *ionosdk.APIResponse, error) {
+	return ac.ComputeClient.UserManagementApi.UmUsersPut(ctx, id).User(u).Execute()
+}
+
+// DeleteUser deletes a user.
+func (ac *APIClient) DeleteUser(ctx context.Context, id string) (*ionosdk.APIResponse, error) {
+	return ac.ComputeClient.UserManagementApi.UmUsersDelete(ctx, id).Execute()
+}
+
+// GetAPIClient returns the ionoscloud APIClient
+func (ac *APIClient) GetAPIClient() *ionosdk.APIClient {
+	return ac.ComputeClient
+}
+
+// IsUserUpToDate returns true if the User is up-to-date or false otherwise.
+func IsUserUpToDate(cr *v1alpha1.User, observed ionosdk.User) bool { // nolint:gocyclo
+	if !observed.HasProperties() || cr == nil {
+		return false
+	}
+
+	// After creation the password is stored as a connection detail secret
+	// and removed from the cr. If the cr has a password it means
+	// the client wants to update it.
+	if cr.Spec.ForProvider.Password != "" {
+		return false
+	}
+
+	props := observed.GetProperties()
+	switch {
+	case cr.Spec.ForProvider.Administrator != *props.GetAdministrator():
+		return false
+	case cr.Spec.ForProvider.Email != *props.GetEmail():
+		return false
+	case cr.Spec.ForProvider.FirstName != *props.GetFirstname():
+		return false
+	case cr.Spec.ForProvider.ForceSecAuth != *props.GetForceSecAuth():
+		return false
+	case cr.Spec.ForProvider.LastName != *props.GetLastname():
+		return false
+	case cr.Spec.ForProvider.SecAuthActive != *props.GetSecAuthActive():
+		return false
+	case cr.Spec.ForProvider.Active != *props.GetActive():
+		return false
+	}
+	return true
+}
+
+type userPropsSetter interface {
```

Comment:
This interface introduces strong coupling with UserPropertiesPost and UserPropertiesPut. Our team has no control over UserPropertiesPost or UserPropertiesPut, and you are forcing them to implement userPropsSetter.

Reply:
Can you explain where is the strong coupling? Because the intention of the interface is exactly the opposite. We have the interface thus we control that. I have added a compile type check to make this even safer.

A bit more of context:
We create and update users. Both has the same set of properties however they are of different types on the sdk. Instead of having a duplicated function for each call, create(UserPropertiesPost) and update(UserPropertiesPut) this interface allow us to have one function to rule them all.

If you think this can be improved, please give a concrete example.

Reply:
when you introduce the interface you build an inheritance relationship. Inheritance is tightly coupled. The UserPropertiesPost and UserPropertiesPut are completely different objects, even they may have similar properties. The fix of this(similar properties) should be the SDK responsibility, and it could be done by composition not inheritance.

func setUserPropertiesPost(p v1alpha1.UserParameters, props UserPropertiesPost) { props.SetFirstname(p.FirstName) ...} and
func setUserPropertiesPut(p v1alpha1.UserParameters, props userPropsSetter) {props.SetFirstname(p.FirstName) ...} 
does not introduce duplication because they are dealing with completely different inputs.

My recommendation is to use two different methods instead of introducing coupling. 

I believe this can end with an arguable discussion, thus I will leave it to you to decide. If you still did not get my point, then we can talk about it.

Reply:
I will choose to leave it as is.